File: platform/sdk/swift/Sources/LayerXSDK/NativeCapabilitySet.swift
import Crypto
import Foundation

public enum NativeCapability: Equatable, Sendable {
    case storageRead
    case storageWrite
    case emitEvent
    case call(program: Data)
    case transfer402(asset: Data, to: Data, maximumAmount: UInt128Value)
    case programSpend(ownerProgram: Data, seed: Data, sourceAccount: Data, asset: Data, to: Data, maximumAmount: UInt128Value)
    case receiptRead(receiptDigest: Data)
    case balanceView(account: Data, asset: Data, receiptDigest: Data)
    case sharedStorageRead
    case sharedStorageWrite
}

public enum NativeCapabilitySet {
    public static let maximumGrants = 238
    public static let maximumBytes = 65452
    public static let maximumBalanceViews = 32

    private struct Entry {
        let rank: Int
        let key: [Data]
        let encoded: Data
        let maximum: UInt128Value
        let receipt: Data
    }
    private static func require(_ condition: Bool) throws {
        guard condition else { throw NativeProgramCallError.invalid }
    }
    private static func nonzero(_ value: Data) throws -> Data {
        try require(value.count == 32 && value.contains(where: { $0 != 0 }))
        return Data(value)
    }
    public static func deriveProgramAccount(owner: Data, seed: Data) throws -> Data {
        let program = try nonzero(owner)
        try require(seed.count <= 128)
        var hash = SHA256()
        hash.update(data: Data("LayerX/programs/program-account/v1\0".utf8))
        hash.update(data: program); hash.update(data: word(UInt64(seed.count), 4)); hash.update(data: seed)
        return Data(hash.finalize())
    }
    private static func word(_ value: UInt64, _ length: Int) -> Data {
        Data((0..<length).reversed().map { UInt8(truncatingIfNeeded: value >> ($0 * 8)) })
    }
    private static func entry(_ grant: NativeCapability) throws -> Entry {
        let rank: Int
        var encoded: Data
        let key: [Data]
        var maximum = UInt128Value(high: 0, low: 0)
        var receipt = Data()
        switch grant {
        case .storageRead: rank = 0; encoded = Data([1]); key = []
        case .storageWrite: rank = 1; encoded = Data([2]); key = []
        case .emitEvent: rank = 2; encoded = Data([3]); key = []
        case .call(let program): rank = 3; encoded = Data([4]); key = [try nonzero(program)]
        case .transfer402(let asset, let to, let amount):
            rank = 4; encoded = Data([5]); key = [try nonzero(asset), try nonzero(to)]; maximum = amount
        case .programSpend(let owner, let seed, let source, let asset, let to, let amount):
            rank = 5; encoded = Data([9])
            let program = try nonzero(owner)
            try require(seed.count <= 128)
            try require(deriveProgramAccount(owner: program, seed: seed) == source)
            key = [program, Data(seed), Data(source), try nonzero(asset), try nonzero(to)]; maximum = amount
        case .receiptRead(let digest): rank = 6; encoded = Data([6]); key = [try nonzero(digest)]
        case .balanceView(let account, let asset, let digest):
            rank = 7; encoded = Data([10]); key = [try nonzero(account), try nonzero(asset)]; receipt = try nonzero(digest)
        case .sharedStorageRead: rank = 8; encoded = Data([7]); key = []
        case .sharedStorageWrite: rank = 9; encoded = Data([8]); key = []
        }
        for (index, field) in key.enumerated() {
            if rank == 5 && index == 1 { encoded.append(word(UInt64(field.count), 2)) }
            encoded.append(field)
        }
        if rank == 4 || rank == 5 {
            try require(maximum.high != 0 || maximum.low != 0)
            encoded.append(word(maximum.high, 8)); encoded.append(word(maximum.low, 8))
        }
        encoded.append(receipt)
        return Entry(rank: rank, key: key, encoded: encoded, maximum: maximum, receipt: receipt)
    }
    private static func compare(_ left: Entry, _ right: Entry) -> Int {
        if left.rank != right.rank { return left.rank < right.rank ? -1 : 1 }
        for (first, second) in zip(left.key, right.key) {
            if first != second { return first.lexicographicallyPrecedes(second) ? -1 : 1 }
        }
        return 0
    }
    public static func encode(_ grants: [NativeCapability]) throws -> Data {
        try require(grants.count <= maximumGrants)
        let entries = try grants.map(entry).sorted { compare($0, $1) < 0 }
        try require(entries.filter { $0.rank == 7 }.count <= maximumBalanceViews)
        var output = word(UInt64(entries.count), 2)
        var prior: Entry?
        for value in entries {
            if let prior { try require(compare(prior, value) < 0) }
            output.append(value.encoded); prior = value
        }
        try require(output.count <= maximumBytes)
        return output
    }
    public static func decode(_ encoded: Data) throws -> [NativeCapability] {
        try require(encoded.count >= 2 && encoded.count <= maximumBytes)
        var cursor = Cursor(bytes: Data(encoded)); let count = Int(try cursor.word(2))
        try require(count <= maximumGrants)
        var grants: [NativeCapability] = []
        for _ in 0..<count {
            let grant: NativeCapability
            switch try cursor.word(1) {
            case 1: grant = .storageRead
            case 2: grant = .storageWrite
            case 3: grant = .emitEvent
            case 4: grant = .call(program: try cursor.take(32))
            case 5: grant = .transfer402(asset: try cursor.take(32), to: try cursor.take(32), maximumAmount: try cursor.amount())
            case 6: grant = .receiptRead(receiptDigest: try cursor.take(32))
            case 7: grant = .sharedStorageRead
            case 8: grant = .sharedStorageWrite
            case 9:
                let owner = try cursor.take(32); let length = Int(try cursor.word(2)); try require(length <= 128)
                grant = .programSpend(ownerProgram: owner, seed: try cursor.take(length), sourceAccount: try cursor.take(32), asset: try cursor.take(32), to: try cursor.take(32), maximumAmount: try cursor.amount())
            case 10: grant = .balanceView(account: try cursor.take(32), asset: try cursor.take(32), receiptDigest: try cursor.take(32))
            default: throw NativeProgramCallError.invalid
            }
            grants.append(grant)
        }
        try require(cursor.offset == encoded.count && encode(grants) == encoded)
        return grants
    }
    public static func narrow(_ parent: [NativeCapability], to requested: [NativeCapability]) throws -> [NativeCapability] {
        let parents = try decode(encode(parent)).map(entry)
        let children = try decode(encode(requested))
        for child in children {
            let value = try entry(child)
            guard let ancestor = parents.first(where: { compare($0, value) == 0 }) else { throw NativeProgramCallError.invalid }
            try require(value.maximum.high < ancestor.maximum.high || value.maximum.high == ancestor.maximum.high && value.maximum.low <= ancestor.maximum.low)
            try require(value.receipt == ancestor.receipt)
        }
        return children
    }
    private struct Cursor {
        let bytes: Data
        var offset = 0
        mutating func take(_ length: Int) throws -> Data {
            try require(length >= 0 && length <= bytes.count - offset)
            let value = bytes.subdata(in: offset..<(offset + length)); offset += length; return value
        }
        mutating func word(_ length: Int) throws -> UInt64 {
            try take(length).reduce(0) { ($0 << 8) | UInt64($1) }
        }
        mutating func amount() throws -> UInt128Value {
            UInt128Value(high: try word(8), low: try word(8))
        }
    }
}

File: platform/sdk/jvm/src/main/java/com/sidiora/layerx/sdk/NativeCapabilitySet.java
package com.sidiora.layerx.sdk;

import java.io.ByteArrayOutputStream;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class NativeCapabilitySet {
    public static final int MAXIMUM_GRANTS = 238;
    public static final int MAXIMUM_BYTES = 65452;
    public static final int MAXIMUM_BALANCE_VIEWS = 32;

    public sealed interface Capability permits StorageRead, StorageWrite, EmitEvent, Call,
            Transfer402, ProgramSpend, ReceiptRead, BalanceView, SharedStorageRead, SharedStorageWrite {}
    public record StorageRead() implements Capability {}
    public record StorageWrite() implements Capability {}
    public record EmitEvent() implements Capability {}
    public record Call(byte[] program) implements Capability {}
    public record Transfer402(byte[] asset, byte[] to, BigInteger maximumAmount) implements Capability {}
    public record ProgramSpend(byte[] ownerProgram, byte[] seed, byte[] sourceAccount, byte[] asset,
                               byte[] to, BigInteger maximumAmount) implements Capability {}
    public record ReceiptRead(byte[] receiptDigest) implements Capability {}
    public record BalanceView(byte[] account, byte[] asset, byte[] receiptDigest) implements Capability {}
    public record SharedStorageRead() implements Capability {}
    public record SharedStorageWrite() implements Capability {}

    private record Entry(int rank, List<byte[]> key, byte[] encoded, BigInteger maximum, byte[] receipt) {}
    private NativeCapabilitySet() {}
    private static void require(boolean condition) {
        if (!condition) throw new IllegalArgumentException("Noncanonical native capability set");
    }
    private static byte[] nonzero(byte[] value) {
        require(value != null && value.length == 32 && !Arrays.equals(value, new byte[32]));
        return value.clone();
    }
    public static byte[] deriveProgramAccount(byte[] owner, byte[] seed) {
        byte[] program = nonzero(owner);
        require(seed != null && seed.length <= 128);
        try {
            MessageDigest hash = MessageDigest.getInstance("SHA-256");
            hash.update("LayerX/programs/program-account/v1\0".getBytes(StandardCharsets.UTF_8));
            hash.update(program); hash.update(ByteBuffer.allocate(4).putInt(seed.length).array());
            return hash.digest(seed);
        } catch (NoSuchAlgorithmException error) { throw new IllegalStateException(error); }
    }
    private static byte[] amount(BigInteger value) {
        require(value != null && value.signum() > 0 && value.bitLength() <= 128);
        byte[] source = value.toByteArray(); byte[] result = new byte[16];
        int length = Math.min(source.length, 16);
        System.arraycopy(source, source.length - length, result, 16 - length, length);
        return result;
    }
    private static Entry entry(Capability grant) {
        require(grant != null);
        var output = new ByteArrayOutputStream(); var key = new ArrayList<byte[]>();
        int rank; BigInteger maximum = BigInteger.ZERO; byte[] receipt = new byte[0];
        if (grant instanceof StorageRead) { rank = 0; output.write(1); }
        else if (grant instanceof StorageWrite) { rank = 1; output.write(2); }
        else if (grant instanceof EmitEvent) { rank = 2; output.write(3); }
        else if (grant instanceof Call value) { rank = 3; output.write(4); key.add(nonzero(value.program())); }
        else if (grant instanceof Transfer402 value) {
            rank = 4; output.write(5); key.add(nonzero(value.asset())); key.add(nonzero(value.to())); maximum = value.maximumAmount();
        } else if (grant instanceof ProgramSpend value) {
            rank = 5; output.write(9);
            byte[] owner = nonzero(value.ownerProgram());
            require(value.seed() != null && value.seed().length <= 128);
            byte[] seed = value.seed().clone();
            require(Arrays.equals(deriveProgramAccount(owner, seed), value.sourceAccount()));
            key.add(owner); key.add(seed); key.add(value.sourceAccount().clone());
            key.add(nonzero(value.asset())); key.add(nonzero(value.to())); maximum = value.maximumAmount();
        } else if (grant instanceof ReceiptRead value) { rank = 6; output.write(6); key.add(nonzero(value.receiptDigest())); }
        else if (grant instanceof BalanceView value) {
            rank = 7; output.write(10); key.add(nonzero(value.account())); key.add(nonzero(value.asset())); receipt = nonzero(value.receiptDigest());
        } else if (grant instanceof SharedStorageRead) { rank = 8; output.write(7); }
        else if (grant instanceof SharedStorageWrite) { rank = 9; output.write(8); }
        else throw new IllegalArgumentException("Unknown native capability");
        for (int index = 0; index < key.size(); index++) {
            byte[] field = key.get(index);
            if (rank == 5 && index == 1) { output.write(field.length >>> 8); output.write(field.length); }
            output.writeBytes(field);
        }
        if (rank == 4 || rank == 5) output.writeBytes(amount(maximum));
        output.writeBytes(receipt);
        return new Entry(rank, key, output.toByteArray(), maximum, receipt);
    }
    private static int compare(Entry left, Entry right) {
        int rank = Integer.compare(left.rank(), right.rank());
        if (rank != 0) return rank;
        for (int index = 0; index < left.key().size(); index++) {
            int order = Arrays.compareUnsigned(left.key().get(index), right.key().get(index));
            if (order != 0) return order;
        }
        return 0;
    }
    public static byte[] encode(List<? extends Capability> grants) {
        require(grants != null && grants.size() <= MAXIMUM_GRANTS);
        var entries = new ArrayList<Entry>(); int balanceViews = 0;
        for (Capability grant : grants) {
            Entry value = entry(grant); entries.add(value);
            if (value.rank() == 7) balanceViews++;
        }
        require(balanceViews <= MAXIMUM_BALANCE_VIEWS);
        entries.sort(NativeCapabilitySet::compare);
        var output = new ByteArrayOutputStream(); output.write(entries.size() >>> 8); output.write(entries.size());
        Entry prior = null;
        for (Entry value : entries) {
            require(prior == null || compare(prior, value) < 0);
            output.writeBytes(value.encoded()); prior = value;
        }
        require(output.size() <= MAXIMUM_BYTES);
        return output.toByteArray();
    }
    private static byte[] take(ByteBuffer cursor, int length) {
        require(length >= 0 && length <= cursor.remaining());
        byte[] value = new byte[length]; cursor.get(value); return value;
    }
    private static int word(ByteBuffer cursor) {
        return Short.toUnsignedInt(ByteBuffer.wrap(take(cursor, 2)).getShort());
    }
    private static BigInteger amount(ByteBuffer cursor) { return new BigInteger(1, take(cursor, 16)); }
    public static List<Capability> decode(byte[] encoded) {
        require(encoded != null && encoded.length >= 2 && encoded.length <= MAXIMUM_BYTES);
        ByteBuffer cursor = ByteBuffer.wrap(encoded); int count = word(cursor);
        require(count <= MAXIMUM_GRANTS); var grants = new ArrayList<Capability>();
        for (int index = 0; index < count; index++) {
            int tag = Byte.toUnsignedInt(take(cursor, 1)[0]);
            Capability grant;
            switch (tag) {
                case 1 -> grant = new StorageRead();
                case 2 -> grant = new StorageWrite();
                case 3 -> grant = new EmitEvent();
                case 4 -> grant = new Call(take(cursor, 32));
                case 5 -> grant = new Transfer402(take(cursor, 32), take(cursor, 32), amount(cursor));
                case 6 -> grant = new ReceiptRead(take(cursor, 32));
                case 7 -> grant = new SharedStorageRead();
                case 8 -> grant = new SharedStorageWrite();
                case 9 -> {
                    byte[] owner = take(cursor, 32); int length = word(cursor); require(length <= 128);
                    grant = new ProgramSpend(owner, take(cursor, length), take(cursor, 32), take(cursor, 32), take(cursor, 32), amount(cursor));
                }
                case 10 -> grant = new BalanceView(take(cursor, 32), take(cursor, 32), take(cursor, 32));
                default -> throw new IllegalArgumentException("Unknown native capability tag");
            }
            grants.add(grant);
        }
        require(!cursor.hasRemaining() && Arrays.equals(encode(grants), encoded));
        return List.copyOf(grants);
    }
    public static List<Capability> narrow(List<? extends Capability> parent, List<? extends Capability> requested) {
        List<Capability> parents = decode(encode(parent)); List<Capability> children = decode(encode(requested));
        for (Capability child : children) {
            Entry value = entry(child); Entry found = null;
            for (Capability grant : parents) {
                Entry ancestor = entry(grant);
                if (compare(ancestor, value) == 0) { found = ancestor; break; }
            }
            require(found != null && value.maximum().compareTo(found.maximum()) <= 0 && Arrays.equals(value.receipt(), found.receipt()));
        }
        return children;
    }
}
